Kernel/Graphics/VirtIOGPU/GraphicsAdapter.cpp

```diff
@@ -422,4 +458,76 @@ void GraphicsAdapter::delete_resource(ResourceID resource_id)
     VERIFY(response.type == static_cast<u32>(Protocol::CommandType::VIRTIO_GPU_RESP_OK_NODATA));
 }
 
+void GraphicsAdapter::initialize_3d_device()
+{
+    if (m_has_virgl_support) {
+        MutexLocker locker(m_operation_lock);
+        m_3d_device = MUST(DeviceManagement::try_create_device<VirtIOGPU::GPU3DDevice>(*this));
+    }
+}
+
+ContextID GraphicsAdapter::create_context()
+{
+    VERIFY(m_operation_lock.is_locked());
+    auto ctx_id = allocate_context_id();
+    auto writer = create_scratchspace_writer();
+    auto& request = writer.append_structure<Protocol::ContextCreate>();
+    auto& response = writer.append_structure<Protocol::ControlHeader>();
+
+    constexpr char const* region_name = "Serenity VirGL3D Context";
+    populate_virtio_gpu_request_header(request.header, Protocol::CommandType::VIRTIO_GPU_CMD_CTX_CREATE, VIRTIO_GPU_FLAG_FENCE);
+    request.header.context_id = ctx_id.value();
+    request.name_length = strlen(region_name);
+    memset(request.debug_name.data(), 0, 64);
+    VERIFY(request.name_length <= 64);
+    memcpy(request.debug_name.data(), region_name, request.name_length);
+
+    synchronous_virtio_gpu_command(start_of_scratch_space(), sizeof(request), sizeof(response));
+
+    VERIFY(response.type == static_cast<u32>(Protocol::CommandType::VIRTIO_GPU_RESP_OK_NODATA));
+    return ctx_id;
+}
+
+void GraphicsAdapter::submit_command_buffer(ContextID context_id, Function<size_t(Bytes)> buffer_writer)
+{
+    VERIFY(m_operation_lock.is_locked());
+    auto writer = create_scratchspace_writer();
+    auto& request = writer.append_structure<Protocol::CommandSubmit>();
+
+    populate_virtio_gpu_request_header(request.header, Protocol::CommandType::VIRTIO_GPU_CMD_SUBMIT_3D, VIRTIO_GPU_FLAG_FENCE);
+    request.header.context_id = context_id.value();
+
+    auto max_command_buffer_length = m_scratch_space->size() - sizeof(request) - sizeof(Protocol::ControlHeader);
+    Bytes command_buffer_buffer(m_scratch_space->vaddr().offset(sizeof(request)).as_ptr(), max_command_buffer_length);
+    request.size = buffer_writer(command_buffer_buffer);
+    writer.skip_bytes(request.size);
+    dbgln_if(VIRTIO_DEBUG, "VirtIO::GraphicsAdapter: Sending command buffer of length {}", request.size);
+    // On x86_64, we may end up with a misaligned ControlHeader for some command buffers
```

Comment:
FIXME?

Reply:
Oh yeah, that's been fixed actually. I think I should just remove that comment entirely.

Reply:
Will remove in next revision.

Reply:
Actually, this loop can be removed as well, if we alignas the controlheader struct.

Reply:
Nvm, the loop cannot be removed, because apparently alignas directives that reduce alignment are silently ignored. I've added a better loop and a better comment.